src/mr_allocator.rs

```diff
@@ -1,60 +1,603 @@
+                (MALLOCX_ALIGN(layout.align()) | MALLOCX_ARENA(self.arena_ind.cast())).cast(),
+            )
+        };
+        if addr.is_null() {
+            None
+        } else {
+            Some(addr.cast::<u8>())
+        }
+    }
+
+    /// Look up `raw_mr` info by addr
+    #[allow(clippy::unreachable)]
+    fn lookup_raw_mr(&self, addr: usize) -> Arc<RawMemoryRegion> {
+        EXTENT_TOKEN_MAP.map_read(|map| {
+            map.range(..addr.overflow_add(1)).next_back().map_or_else(
+                || {
+                    unreachable!("can not find raw mr by addr {}", addr);
+                },
+                |(_, item)| {
+                    debug!("LOOK addr {}, item {:?}", addr, item);
+                    assert_eq!(self.arena_ind, item.arena_ind);
+                    assert!(addr >= item.addr && addr < item.addr.overflow_add(item.len));
+                    Arc::<RawMemoryRegion>::clone(&item.raw_mr)
+                },
+            )
+        })
+    }
+}
+
+/// Parking lot mutex wrapper.
+/// Provides functional expression methods.
+pub(crate) struct ParkingLotMutex<T> {
+    /// Parking_lot mutex
+    inner: Mutex<T>,
+}
+
+impl<T> ParkingLotMutex<T> {
+    /// New a `ParkingLotMutex`
+    fn new(v: T) -> Self {
+        Self {
+            inner: Mutex::new(v),
+        }
+    }
+
+    /// Use `func` to read the value in the mutex
+    fn map_read<F, R>(&self, func: F) -> R
+    where
+        F: FnOnce(&MutexGuard<'_, T>) -> R,
+    {
+        let mut guard = self.inner.try_lock();
+        while guard.is_none() {
+            guard = self.inner.try_lock();
+        }
+        // guard is checked
+        #[allow(clippy::unwrap_used)]
+        func(&guard.unwrap())
+    }
+
+    /// Use `func` to write the value in the mutex
+    fn map_write<F, R>(&self, func: F) -> R
+    where
+        F: FnOnce(&mut MutexGuard<'_, T>) -> R,
+    {
+        let mut guard = self.inner.try_lock();
+        while guard.is_none() {
+            guard = self.inner.try_lock();
+        }
+        // guard is checked
+        #[allow(clippy::unwrap_used)]
+        func(&mut guard.unwrap())
+    }
+}
+
+/// Get default extent hooks of jemalloc
+#[allow(clippy::unreachable)]
+fn get_default_hooks_impl(arena_ind: u32) -> extent_hooks_t {
+    // read default alloc impl
+    let mut hooks: *mut extent_hooks_t = ptr::null_mut();
+    let hooks_ptr: *mut *mut extent_hooks_t = &mut hooks;
+    let key = format!("arena.{}.extent_hooks\0", arena_ind);
+    let mut hooks_len = mem::size_of_val(&hooks);
+    let errno = unsafe {
+        tikv_jemalloc_sys::mallctl(
+            key.as_ptr().cast(),
+            hooks_ptr.cast(),
+            &mut hooks_len,
+            ptr::null_mut(),
+            0,
+        )
+    };
+    let hooksd = unsafe { *hooks };
+    if errno != 0_i32 {
+        unreachable!(
+            "get default hooks failed {:?}",
+            io::Error::from_raw_os_error(errno)
+        );
+    }
+    hooksd
```

Comment:
`clippy::undocumented_unsafe_blocks`
Does the `hooks` pointer is valid even if an error occurs?

Reply:
It is better to check errno first : )